rfcs/config-luaurc.md

```diff
@@ -0,0 +1,66 @@
+# Configure analysis via .luaurc
+
+## Summary
+
+Introduces a way to configure type checker and linter using JSON-like .luaurc files
+
+## Motivation
+
+While Luau analysis tools try to provide sensible defaults, it's difficult to establish the rules that work for all code.
+For example, some packages may decide that unused variables aren't interesting; other packages may decide that all files should be using strict typechecking mode.
+
+While it's possible to configure some aspects of analysis behavior using --! comments, it can be cumbersome to replicate this in all files.
+
+## Design
+
+To solve this problem, we are going to introduce support for `.luaurc` files for users of command-line Luau tools.
+For a given .lua file, Luau will search for .luaurc files starting from the folder that the .lua file is in; all files in the ancestry chain will be parsed and their configuration
+applied. When multiple files are used, the file closer to the .lua file overrides the settings.
+
+.luaurc is a JSON file that can also contain comments and trailing commas. The file can have the following keys:
+
+- `"mode"`: type checking mode, can be one of "nocheck", "nonstrict", "strict"
+- `"lint"`: lints to enable; points to an array of string literals that correspond to the names of linting rules (see https://luau-lang.org/lint), or `"*"` that means "all rules"
+- `"nolint"`: lints to disable; points to an array of string literals or `"*"`
+- `"lintErrors"`: a boolean that controls whether lint issues are reported as errors or warnings (off by default)
+- `"typeErrors"`: a boolean that controls whether type issues are reported as errors or warnings (on by default)
+- `"globals"`: extra global values; points to an array of strings where each string names a global that the type checker and linter must assume is valid and of type `any`
```

Comment:
Do we need this if we also have definition files?

Reply:
Maybe not, however I've intentionally wanted something simple that covers many use cases and doesn't include the def files - today the def files grammar isn't stable, and also referring to separate files presents extra challenges when you consider non-file-base workloads...

What I think would make sense as a next step for def files is to support them as command line inputs and make the grammar "official", and only then start incorporating them into the path-based configs.

Reply:
Fair. Once we stabilize definition files, I assume we'd deprecate and eventually remove this in favor of a list of paths to definition files.

Reply:
Yeah or really another way to think about it is as layering - this is a concise way to specify extra APIs that the type checker / linter can't know about, a next step is to specify a defn file where every member is `any`, a next step is to detail the type. So it's not even necessarily the case that we'd deprecate this, it's just going to be a more lightweight less precise way to talk about things analysis doesn't know about.